nbconvert/filters/strings.py

```diff
@@ -188,14 +188,59 @@ def get_lines(text, start=None,end=None):
     # Return the right lines.
     return "\n".join(lines[start:end]) #re-join
 
-def ipython2python(code):
+def _mpl_magic_regex_generator(forbidden_backends=None):
+    """Create regex for detecting incompatible backends in mpl magics.
+    
+    Parameters
+    ----------
+
+    forbidden_backens: List
+        List of strings, each of which is a forbidden matplotlib backend. 
+    """
+    if forbidden_backends is None:
+        forbidden_backends = ["notebook", "inline"]
+    no_mpl_be_str = "|".join(forbidden_backends)
+    mpl_magic_backend_extractor = re.compile(r'(^%matplotlib \w*)({})'.format(no_mpl_be_str))
+    return mpl_magic_backend_extractor
+
+mpl_magic_regex = _mpl_magic_regex_generator()
+
+def _remove_gui_mpl_backends(code, bad_backends=None):
+    """Remove incompatible backends from mpl magics.
+    
+    Parameters
+    ----------
+
+    code : str
+        IPython code with matplotlib magic
+    bad_backends: list of str
+        List of strings, each of which is a forbidden matplotlib backend.
+    """
+    if bad_backends is not None:
+        mpl_magic_regex_local = _mpl_magic_regex_generator(bad_backends)
+    else: 
+        mpl_magic_regex_local = mpl_magic_regex
+
+    lines = code.split('\n',1)
+    notebook_magic = mpl_magic_regex_local.match(lines[0])
```

Comment:
Since magic can occur anywhere in a cell this would miss cases like,

```python
pass
%matplotlib inline
```
We could check every line, but then we have to grapple with multiline strings.